src/utils.rs

```diff
@@ -0,0 +1,130 @@
+use scale_info::{form::PortableForm, Field, PortableRegistry, Type, TypeDef, TypeParameter};
+use smallvec::{smallvec, SmallVec};
+use std::collections::HashMap;
+
+pub fn ensure_unique_type_paths(types: &mut PortableRegistry) {
+    let mut types_with_same_type_path = HashMap::<&[String], SmallVec<[u32; 2]>>::new();
+
+    // Figure out which types have the same type paths by storing them in the HashMap
+    for ty in types.types.iter() {
+        // Ignore types without a path (i.e prelude types).
+        if ty.ty.path.namespace().is_empty() {
+            continue;
+        };
+        types_with_same_type_path
+            .entry(&ty.ty.path.segments)
+            .or_default()
+            .push(ty.id);
+    }
+
+    let clashing_type_ids = types_with_same_type_path
+        .into_iter()
+        .filter_map(|(_, v)| (v.len() > 1).then_some(v));
+
+    // submit to this buffer type ids (u32), where a number (usize) should be added to the type path to distinguish them.
+    let mut renaming_commands: Vec<(u32, usize)> = vec![];
+
+    for type_ids in clashing_type_ids {
+        // Map N types in type_ids to M new type paths, where M <= N.
+        // types with the same structure should map to the same type path after all.
+        let mut types_with_same_structure: Vec<(&Type<PortableForm>, SmallVec<[u32; 2]>)> = vec![];
```

Comment:
We coud have 4 types on the same path, named A, B, C, D.
And let's say that:
- types_equal_extended_to_params(A, C) == true
- types_equal_extended_to_params(B, D) == true
- types_equal_extended_to_params(A, B) == false

Does this still deduplicate/rename all of them?

Reply:
I think in this case, we iterate over A, B, C, D:
```
- insert A at index 0,
- insert B at index 1,
- insert C at index 0, because it equals A,
- insert D at index 1, because it equals B but not A.
```
So A and C will get `PathName1` and B and D get `PathName2`. Which should be deduplicated enough.